trio/_core/tests/test_guest_mode.py

```diff
@@ -0,0 +1,361 @@
+import pytest
+import asyncio
+import traceback
+import queue
+from functools import partial
+from math import inf
+import signal
+import socket
+
+import trio
+import trio.testing
+from .tutil import gc_collect_harder
+
+# The simplest possible "host" loop.
+# Nice features:
+# - we can run code "outside" of trio using the schedule function passed to
+#   our main
+# - final result is returned
+# - any unhandled exceptions cause an immediate crash
+def trivial_guest_run(trio_fn, **start_guest_run_kwargs):
+    todo = queue.Queue()
+
+    def run_sync_soon_threadsafe(fn):
+        todo.put(("run", fn))
+
+    def done_callback(outcome):
+        todo.put(("unwrap", outcome))
+
+    trio.lowlevel.start_guest_run(
+        trio_fn,
+        run_sync_soon_threadsafe,
+        run_sync_soon_threadsafe=run_sync_soon_threadsafe,
+        done_callback=done_callback,
+        **start_guest_run_kwargs,
+    )
+
+    try:
+        while True:
+            op, obj = todo.get()
+            if op == "run":
+                obj()
+            elif op == "unwrap":
+                return obj.unwrap()
+            else:  # pragma: no cover
+                assert False
+    finally:
+        # Make sure that exceptions raised here don't capture these, so that
+        # if an exception does cause us to abandon a run then the Trio state
+        # has a chance to be GC'ed and warn about it.
+        del todo, run_sync_soon_threadsafe, done_callback
+
+
+def test_guest_trivial():
+    async def trio_return(in_host):
+        await trio.sleep(0)
+        return "ok"
+
+    assert trivial_guest_run(trio_return) == "ok"
+
+    async def trio_fail(in_host):
+        raise KeyError("whoopsiedaisy")
+
+    with pytest.raises(KeyError, match="whoopsiedaisy"):
+        trivial_guest_run(trio_fail)
+
+
+def test_guest_can_do_io():
+    async def trio_main(in_host):
+        record = []
+        a, b = trio.socket.socketpair()
+        with a, b:
+            async with trio.open_nursery() as nursery:
+
+                async def do_receive():
+                    record.append(await a.recv(1))
+
+                nursery.start_soon(do_receive)
+                await trio.testing.wait_all_tasks_blocked()
+
+                await b.send(b"x")
+
+        assert record == [b"x"]
+
+    trivial_guest_run(trio_main)
+
+
+def test_host_can_directly_wake_trio_task():
+    async def trio_main(in_host):
+        ev = trio.Event()
+        in_host(ev.set)
+        await ev.wait()
+        return "ok"
+
+    assert trivial_guest_run(trio_main) == "ok"
+
+
+def test_host_altering_deadlines_wakes_trio_up():
+    def set_deadline(cscope, new_deadline):
+        cscope.deadline = new_deadline
+
+    async def trio_main(in_host):
+        with trio.CancelScope() as cscope:
+            in_host(lambda: set_deadline(cscope, -inf))
+            await trio.sleep_forever()
+            assert cscope.cancelled_caught
```

Comment:
I think these asserts should be dedented (AFAICT you're currently skipping them when the cancellation is raised on the previous line)

Reply:
Nice catch!